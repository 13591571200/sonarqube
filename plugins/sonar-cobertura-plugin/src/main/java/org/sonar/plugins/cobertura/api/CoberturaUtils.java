package org.sonar.plugins.cobertura.api;

import org.sonar.api.CoreProperties;
import org.sonar.api.batch.maven.MavenPlugin;
import org.sonar.api.resources.Project;
import org.sonar.api.utils.Logs;
import org.sonar.plugins.cobertura.CoberturaMavenPluginHandler;

import java.io.File;

/**
 * @since 2.4
 */
public final class CoberturaUtils {

  public static File getReport(Project project) {
    File report = getReportFromProperty(project);
    if (report == null) {
      report = getReportFromPluginConfiguration(project);
    }
    if (report == null) {
      report = getReportFromDefaultPath(project);
    }

    if (report == null || !report.exists() || !report.isFile()) {
      Logs.INFO.warn("Cobertura report not found at {}", report);
      report = null;
    }
    return report;
  }

  private static File getReportFromProperty(Project project) {
    String path = (String) project.getProperty(CoreProperties.COBERTURA_REPORT_PATH_PROPERTY);
    if (path != null) {
      return project.getFileSystem().resolvePath(path);
    }
    return null;
  }

  private static File getReportFromPluginConfiguration(Project project) {
    MavenPlugin mavenPlugin = MavenPlugin.getPlugin(project.getPom(), CoberturaMavenPluginHandler.GROUP_ID, CoberturaMavenPluginHandler.ARTIFACT_ID);
    if (mavenPlugin != null) {
      String path = mavenPlugin.getParameter("outputDirectory");
      if (path != null) {
        return new File(project.getFileSystem().resolvePath(path), "coverage.xml");
      }
    }
    return null;
  }

  private static File getReportFromDefaultPath(Project project) {
    return new File(project.getFileSystem().getReportOutputDir(), "cobertura/coverage.xml");
  }

  private CoberturaUtils() {
  }

}
