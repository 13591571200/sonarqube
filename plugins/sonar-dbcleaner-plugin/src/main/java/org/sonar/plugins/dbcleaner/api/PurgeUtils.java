/*
 * Sonar, open source software quality management tool.
 * Copyright (C) 2008-2012 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * Sonar is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Sonar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Sonar; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package org.sonar.plugins.dbcleaner.api;

import org.apache.commons.configuration.Configuration;
import org.sonar.api.batch.Event;
import org.sonar.api.database.DatabaseSession;
import org.sonar.api.database.model.*;
import org.sonar.api.design.DependencyDto;
import org.sonar.api.utils.TimeProfiler;

import com.google.common.annotations.VisibleForTesting;

import javax.persistence.Query;
import java.util.List;

/**
 * @since 2.5
 */
public final class PurgeUtils {

  public static final int DEFAULT_MINIMUM_PERIOD_IN_HOURS = 12;
  public static final String PROP_KEY_MINIMUM_PERIOD_IN_HOURS = "sonar.purge.minimumPeriodInHours";

  /**
   * Maximum elements in the SQL statement "IN" due to an Oracle limitation (see error ORA-01795)
   */
  public static final int MAX_IN_ELEMENTS = 950;

  private PurgeUtils() {
    // only static methods
  }

  public static int getMinimumPeriodInHours(Configuration conf) {
    int hours = DEFAULT_MINIMUM_PERIOD_IN_HOURS;
    if (conf != null) {
      hours = conf.getInt(PROP_KEY_MINIMUM_PERIOD_IN_HOURS, DEFAULT_MINIMUM_PERIOD_IN_HOURS);
    }
    return hours;
  }

  public static void deleteSnapshotsData(DatabaseSession session, List<Integer> snapshotIds) {
    deleteMeasuresBySnapshotId(session, snapshotIds);
    deleteSources(session, snapshotIds);
    deleteViolations(session, snapshotIds);
    deleteDependencies(session, snapshotIds);
    deleteDuplicationBlocks(session, snapshotIds);
    deleteEvents(session, snapshotIds);
    deleteSnapshots(session, snapshotIds);
  }

  public static void deleteDependencies(DatabaseSession session, List<Integer> snapshotIds) {
    executeQuery(session, "delete dependencies", snapshotIds, "delete from " + DependencyDto.class.getSimpleName() + " d where d.fromSnapshotId in (:ids)");
    executeQuery(session, "delete dependencies", snapshotIds, "delete from " + DependencyDto.class.getSimpleName() + " d where d.toSnapshotId in (:ids)");
  }

  /**
   * Delete all measures, including MEASURE_DATA
   */
  public static void deleteMeasuresBySnapshotId(DatabaseSession session, List<Integer> snapshotIds) {
    executeQuery(session, "delete measures by snapshot id", snapshotIds, "delete from " + MeasureData.class.getSimpleName() + " m where m.snapshotId in (:ids)");
    executeQuery(session, "delete measures by snapshot id", snapshotIds, "delete from " + MeasureModel.class.getSimpleName() + " m where m.snapshotId in (:ids)");
  }

  /**
   * Delete all measures, including MEASURE_DATA
   */
  public static void deleteMeasuresById(DatabaseSession session, List<Integer> measureIds) {
    executeQuery(session, "delete measures by id", measureIds, "delete from " + MeasureData.class.getSimpleName() + " m where m.measure.id in (:ids)");
    executeQuery(session, "delete measures by id", measureIds, "delete from " + MeasureModel.class.getSimpleName() + " m where m.id in (:ids)");
  }

  /**
   * Delete SNAPSHOT_SOURCES table
   */
  public static void deleteSources(DatabaseSession session, List<Integer> snapshotIds) {
    executeQuery(session, "delete sources", snapshotIds, "delete from " + SnapshotSource.class.getSimpleName() + " e where e.snapshotId in (:ids)");
  }

  /**
   * Delete violations (RULE_FAILURES table)
   */
  public static void deleteViolations(DatabaseSession session, List<Integer> snapshotIds) {
    executeQuery(session, "delete violations", snapshotIds, "delete from " + RuleFailureModel.class.getSimpleName() + " e where e.snapshotId in (:ids)");
  }

  /**
   * Delete DUPLICATIONS_INDEX table
   *
   * @since 2.11
   */
  private static void deleteDuplicationBlocks(DatabaseSession session, List<Integer> snapshotIds) {
    executeNativeQuery(session, "delete duplication blocks", snapshotIds, "delete from duplications_index where snapshot_id in (:ids)");
  }

  /**
   * Delete EVENTS table
   */
  public static void deleteEvents(DatabaseSession session, List<Integer> snapshotIds) {
    executeQuery(session, "delete events", snapshotIds, "delete from " + Event.class.getSimpleName() + " e where e.snapshot.id in (:ids)");
  }

  /**
   * Delete SNAPSHOTS table
   */
  public static void deleteSnapshots(DatabaseSession session, List<Integer> snapshotIds) {
    executeQuery(session, "delete snapshots", snapshotIds, "delete from " + Snapshot.class.getSimpleName() + " s where s.id in (:ids)");
  }

  public static void deleteResources(DatabaseSession session, List<Integer> ids) {
    executeQuery(session, "", ids, "DELETE FROM " + ResourceModel.class.getSimpleName() + " WHERE id in (:ids)");
    deleteResourceIndex(session, ids);
  }

  /**
   * Delete RESOURCE_INDEX table
   */
  public static void deleteResourceIndex(DatabaseSession session, List<Integer> resourceIds) {
    executeNativeQuery(session, "delete resource_index", resourceIds, "delete from resource_index where resource_id in (:ids)");
  }

  /**
   * Paginate execution of SQL requests to avoid exceeding size of rollback segment
   */
  public static void executeQuery(DatabaseSession session, String description, List<Integer> ids, String hql) {
    if (ids == null || ids.isEmpty()) {
      return;
    }
    TimeProfiler profiler = new TimeProfiler().setLevelToDebug().start("Execute " + description);
    int index = 0;
    while (index < ids.size()) {
      List<Integer> paginedSids = ids.subList(index, Math.min(ids.size(), index + MAX_IN_ELEMENTS));
      Query query = session.createQuery(hql);
      query.setParameter("ids", paginedSids);
      query.executeUpdate();
      index += MAX_IN_ELEMENTS;
      session.commit();
    }
    profiler.stop();
  }

  /**
   * @since 2.13
   */
  @VisibleForTesting
  static void executeNativeQuery(DatabaseSession session, String description, List<Integer> ids, String sql) {
    if (ids == null || ids.isEmpty()) {
      return;
    }
    TimeProfiler profiler = new TimeProfiler().setLevelToDebug().start("Execute " + description);
    int index = 0;
    while (index < ids.size()) {
      List<Integer> paginedSids = ids.subList(index, Math.min(ids.size(), index + MAX_IN_ELEMENTS));
      Query query = session.createNativeQuery(sql);
      query.setParameter("ids", paginedSids);
      query.executeUpdate();
      index += MAX_IN_ELEMENTS;
      session.commit();
    }
    profiler.stop();
  }

}
