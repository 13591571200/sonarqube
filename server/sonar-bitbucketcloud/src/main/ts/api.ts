/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { parseError, parseJSON, request, RequestData } from '@sqcore/helpers/request';
import { Paging, MyProject, IdentityProvider } from '@sqcore/app/types';
import {
  AppContext,
  ProjectData,
  RepositoryData,
  RepositoryMeasure,
  PullRequestData,
  PullRequestContext
} from './types';
import { displayMessage, proxyRequest, apiRequest, getAppKey, getRepoUuid } from './utils';

export function bindProject(data: AppContext & { projectKey: string }): Promise<void | Response> {
  return post('/integration/bitbucketcloud/bind_repo', data).catch(displayWSError);
}

export function displayWSError({ response }: { response: Response }): Promise<Response> {
  parseError({ response }).then(
    msg => {
      displayMessage('error', 'SonarCloud Error', msg);
    },
    () => {}
  );
  return Promise.reject(response);
}

export function getIdentityProviders(): Promise<{ identityProviders: IdentityProvider[] }> {
  return getJSON('/api/users/identity_providers');
}

export function getMyProjects(data: {
  p?: number;
  ps?: number;
}): Promise<{ paging: Paging; projects: MyProject[] }> {
  return getJSON('/api/projects/search_my_projects', data);
}

export function getRepositoryData(data: AppContext): Promise<ProjectData> {
  return getJSON('/integration/bitbucketcloud/repo_widget_data', data).then(
    (repoData: RepositoryData) => {
      const measures: { [key: string]: string } = {};
      repoData.measures.forEach((measure: RepositoryMeasure) => {
        if (measure.value !== undefined) {
          measures[measure.metric] = measure.value;
        }
      });
      return { ...repoData, measures };
    }
  );
}

export function getStoredProperty(property: string) {
  return apiRequest({
    cache: false,
    type: 'GET',
    url: `/2.0/repositories/{}/${getRepoUuid()}/properties/${getAppKey()}/${property}`
  });
}

export function putStoredProperty(property: string, value: Object | string | boolean | number) {
  return apiRequest({
    cache: false,
    contentType: 'application/json',
    data: JSON.stringify(value),
    type: 'PUT',
    url: `/2.0/repositories/{}/${getRepoUuid()}/properties/${getAppKey()}/${property}`
  });
}

export function getPullRequestData(context: PullRequestContext): Promise<PullRequestData> {
  return getProxyJSON(
    `/repositories/${context.repoUuid}/pullrequests/${context.prId}/code_quality`
  );
}

function checkStatus(response: Response): Promise<Response> {
  return new Promise((resolve, reject) => {
    if (response.status >= 200 && response.status < 300) {
      resolve(response);
    } else {
      reject({ response });
    }
  });
}

function getProxyJSON(url: string): Promise<any> {
  return proxyRequest({ cache: false, type: 'GET', url });
}

function getJSON(url: string, data?: RequestData): Promise<any> {
  return request(url)
    .setData(data)
    .submit()
    .then(checkStatus)
    .then(parseJSON);
}

function post(url: string, data?: RequestData): Promise<void> {
  return new Promise((resolve, reject) => {
    request(url)
      .setMethod('POST')
      .setData(data)
      .submit()
      .then(checkStatus)
      .then(() => resolve(), reject);
  });
}
