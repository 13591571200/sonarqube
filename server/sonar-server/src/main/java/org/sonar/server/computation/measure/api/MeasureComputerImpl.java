/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.sonar.server.computation.measure.api;

import com.google.common.collect.ImmutableSet;
import java.util.Set;
import org.sonar.api.ce.measure.MeasureComputer;

import static com.google.common.base.Preconditions.checkArgument;

public class MeasureComputerImpl implements MeasureComputer {

  private final Set<String> inputMetricKeys;
  private final Set<String> outputMetrics;
  private final Implementation implementation;

  public MeasureComputerImpl(MeasureComputerBuilderImpl builder) {
    this.inputMetricKeys = ImmutableSet.copyOf(builder.inputMetricKeys);
    this.outputMetrics = ImmutableSet.copyOf(builder.outputMetrics);
    this.implementation = builder.measureComputerImplementation;
  }

  @Override
  public Set<String> getInputMetrics() {
    return inputMetricKeys;
  }

  @Override
  public Set<String> getOutputMetrics() {
    return outputMetrics;
  }

  @Override
  public Implementation getImplementation() {
    return implementation;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    MeasureComputerImpl that = (MeasureComputerImpl) o;

    if (!inputMetricKeys.equals(that.inputMetricKeys)) {
      return false;
    }
    return outputMetrics.equals(that.outputMetrics);
  }

  @Override
  public int hashCode() {
    int result = inputMetricKeys.hashCode();
    result = 31 * result + outputMetrics.hashCode();
    return result;
  }

  @Override
  public String toString() {
    return "MeasureComputerImpl{" +
      "inputMetricKeys=" + inputMetricKeys +
      ", outputMetrics=" + outputMetrics +
      ", implementation=" + implementation +
      '}';
  }

  public static class MeasureComputerBuilderImpl implements MeasureComputerBuilder {

    private String[] inputMetricKeys;
    private String[] outputMetrics;
    private Implementation measureComputerImplementation;

    @Override
    public MeasureComputerBuilder setInputMetrics(String... inputMetrics) {
      checkArgument(inputMetrics != null && inputMetrics.length > 0, "At least one input metric must be defined");
      this.inputMetricKeys = inputMetrics;
      return this;
    }

    @Override
    public MeasureComputerBuilder setOutputMetrics(String... outputMetrics) {
      checkArgument(outputMetrics != null && outputMetrics.length > 0, "At least one output metric must be defined");
      this.outputMetrics = outputMetrics;
      return this;
    }

    @Override
    public MeasureComputerBuilder setImplementation(Implementation impl) {
      checkImplementation(impl);
      this.measureComputerImplementation = impl;
      return this;
    }

    @Override
    public MeasureComputer build() {
      checkArgument(this.inputMetricKeys != null, "At least one input metric must be defined");
      checkArgument(this.outputMetrics != null, "At least one output metric must be defined");
      checkImplementation(this.measureComputerImplementation);
      return new MeasureComputerImpl(this);
    }

    private static void checkImplementation(Implementation impl) {
      checkArgument(impl != null, "The implementation is missing");
    }
  }
}
