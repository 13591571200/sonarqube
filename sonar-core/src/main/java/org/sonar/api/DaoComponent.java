package org.sonar.api;

/**
 * @since 4.4
 */
public interface DaoComponent {
}
