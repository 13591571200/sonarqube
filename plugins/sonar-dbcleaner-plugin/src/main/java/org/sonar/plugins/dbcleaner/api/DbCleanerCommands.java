/*
 * Sonar, open source software quality management tool.
 * Copyright (C) 2008-2011 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * Sonar is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * Sonar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Sonar; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
 */
package org.sonar.plugins.dbcleaner.api;

import org.sonar.api.BatchExtension;
import org.sonar.api.database.DatabaseSession;
import org.sonar.api.database.model.ResourceModel;
import org.sonar.persistence.resource.ResourceIndexerDao;

import java.util.List;

/**
 * @since 2.13
 */
public class DbCleanerCommands implements BatchExtension {

  private DatabaseSession session;
  private ResourceIndexerDao resourceIndexer;

  public DbCleanerCommands(DatabaseSession session, ResourceIndexerDao resourceIndexer) {
    this.session = session;
    this.resourceIndexer = resourceIndexer;
  }

  public DbCleanerCommands deleteSnapshots(List<Integer> snapshotIds, boolean includeDependents) {
    if (includeDependents) {
      PurgeUtils.deleteSnapshotsData(session, snapshotIds);
    } else {
      PurgeUtils.deleteSnapshots(session, snapshotIds);
    }
    return this;
  }

  public DbCleanerCommands deleteResources(List<Integer> resourceIds) {
    PurgeUtils.executeQuery(session, "", resourceIds, "DELETE FROM " + ResourceModel.class.getSimpleName() + " WHERE id in (:ids)");
    resourceIndexer.delete(resourceIds);
    return this;
  }
}
