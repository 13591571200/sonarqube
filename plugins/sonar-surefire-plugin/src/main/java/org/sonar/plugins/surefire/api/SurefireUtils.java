package org.sonar.plugins.surefire.api;

import org.sonar.api.CoreProperties;
import org.sonar.api.batch.maven.MavenPlugin;
import org.sonar.api.batch.maven.MavenSurefireUtils;
import org.sonar.api.resources.Project;

import java.io.File;

/**
 * @since 2.4
 */
public final class SurefireUtils {

  public static File getReportsDirectory(Project project) {
    File dir = getReportsDirectoryFromProperty(project);
    if (dir == null) {
      dir = getReportsDirectoryFromPluginConfiguration(project);
    }
    if (dir == null) {
      dir = getReportsDirectoryFromDefaultConfiguration(project);
    }
    return dir;
  }

  private static File getReportsDirectoryFromProperty(Project project) {
    String path = (String) project.getProperty(CoreProperties.SUREFIRE_REPORTS_PATH_PROPERTY);
    if (path != null) {
      return project.getFileSystem().resolvePath(path);
    }
    return null;
  }

  private static File getReportsDirectoryFromPluginConfiguration(Project project) {
    MavenPlugin plugin = MavenPlugin.getPlugin(project.getPom(), MavenSurefireUtils.GROUP_ID, MavenSurefireUtils.ARTIFACT_ID);
    if (plugin != null) {
      String path = plugin.getParameter("reportsDirectory");
      if (path != null) {
        return project.getFileSystem().resolvePath(path);
      }
    }
    return null;
  }

  private static File getReportsDirectoryFromDefaultConfiguration(Project project) {
    return new File(project.getFileSystem().getBuildDir(), "surefire-reports");
  }

  private SurefireUtils() {
  }

}
