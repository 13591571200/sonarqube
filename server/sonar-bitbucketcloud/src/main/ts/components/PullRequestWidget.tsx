/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Spinner from '@atlaskit/spinner';
import QuestionCircleIcon from '@atlaskit/icon/glyph/question-circle';
import BugIcon from '@sqcore/components/icons-components/BugIcon';
import CodeSmellIcon from '@sqcore/components/icons-components/CodeSmellIcon';
import StatusIndicator from '@sqcore/components/common/StatusIndicator';
import VulnerabilityIcon from '@sqcore/components/icons-components/VulnerabilityIcon';
import { formatMeasure } from '@sqcore/helpers/measures';
import {
  getPullRequestUrl,
  getPathUrlAsString,
  getShortLivingBranchUrl
} from '@sqcore/helpers/urls';
import { getBranchQualityGateColor } from '@sqcore/helpers/branches';
import { getPullRequestData } from '../api';
import { PullRequestData, PullRequestContext } from '../types';
import { getRepoSettingsUrl, isRepoAdmin } from '../utils';

interface Props {
  context: PullRequestContext;
}

interface State {
  error?: string;
  loading: boolean;
  pullRequest?: PullRequestData;
  settingsUrl?: string;
}

export default class PullRequestWidget extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = { loading: true };

  componentDidMount() {
    this.mounted = true;
    this.fetchPullRequestData();
    getRepoSettingsUrl().then(
      settingsUrl => {
        if (this.mounted) {
          this.setState({ settingsUrl });
        }
      },
      () => {}
    );
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchPullRequestData = () => {
    getPullRequestData(this.props.context).then(
      pullRequest => {
        if (this.mounted) {
          this.setState({ error: undefined, loading: false, pullRequest });
        }
      },
      ({ errors }) => {
        if (this.mounted) {
          this.setState({ error: errors && errors[0] && errors[0].msg, loading: false });
        }
      }
    );
  };

  renderStatus = (pullRequest: PullRequestData) => {
    return (
      <span title={`Quality Gate ${formatMeasure(pullRequest.status.qualityGateStatus, 'LEVEL')}`}>
        <StatusIndicator
          className="little-spacer-left little-spacer-right"
          color={getBranchQualityGateColor(pullRequest.status.qualityGateStatus)}
          size="small"
        />
      </span>
    );
  };

  render() {
    const { error, loading, pullRequest, settingsUrl } = this.state;

    if (loading) {
      return (
        <div className="pr-widget">
          <Spinner size="small" />
        </div>
      );
    }

    if (!pullRequest) {
      if (error && error.includes('Repository not bound')) {
        const settingsLink = settingsUrl ? (
          <a href={settingsUrl} rel="noopener noreferrer" target="_parent">
            repository settings
          </a>
        ) : (
          'repository settings'
        );
        return (
          <p className="pr-widget empty">
            {isRepoAdmin() ? (
              <>No link between repository and SonarCloud project. Go to your {settingsLink}.</>
            ) : (
              'No link between repository and SonarCloud project. Contact an administrator.'
            )}
          </p>
        );
      }
      return <p className="pr-widget empty">Not analyzed on SonarCloud yet</p>;
    }

    const totalIssues =
      pullRequest.status.bugs + pullRequest.status.vulnerabilities + pullRequest.status.codeSmells;

    if (totalIssues <= 0) {
      return (
        <p className="pr-widget">
          {this.renderStatus(pullRequest)}
          {"Hooray! SonarCloud analysis didn't find any issue!"}
        </p>
      );
    }

    const shouldDisplayHelper = pullRequest.status.qualityGateStatus === 'OK';

    const prUrl = getPathUrlAsString(
      pullRequest.isPullRequest !== undefined
        ? getPullRequestUrl(pullRequest.projectKey, pullRequest.key)
        : getShortLivingBranchUrl(pullRequest.projectKey, pullRequest.key)
    );

    return (
      <div className="pr-widget">
        <div className="pr-widget-inner">
          {this.renderStatus(pullRequest)}
          <span className="little-spacer-right">SonarCloud found issues:</span>
          <span className="little-spacer-right">
            <BugIcon className="little-spacer-right" />{' '}
            {formatMeasure(pullRequest.status.bugs, 'SHORT_INT')} Bugs
          </span>
          <span className="little-spacer-right">
            <VulnerabilityIcon className="little-spacer-right" />{' '}
            {formatMeasure(pullRequest.status.vulnerabilities, 'SHORT_INT')} Vulnerabilities
          </span>
          <span className="little-spacer-right">
            <CodeSmellIcon className="little-spacer-right" />{' '}
            {formatMeasure(pullRequest.status.codeSmells, 'SHORT_INT')} Code Smells
          </span>
          {shouldDisplayHelper && (
            <span
              title={`The branch Quality Gate is "Passed" because there are no open issue. The remaining ${totalIssues} issue(s) have been confirmed.`}>
              <QuestionCircleIcon label="help" primaryColor="rgb(180, 180, 180)" size="small" />
            </span>
          )}
        </div>
        <a href={prUrl} target="_blank">
          See all issues on SonarCloud
        </a>
      </div>
    );
  }
}
