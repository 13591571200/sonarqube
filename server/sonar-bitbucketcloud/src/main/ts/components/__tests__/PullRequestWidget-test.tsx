/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable camelcase */
import * as React from 'react';
import { shallow } from 'enzyme';
import PullRequestWidget from '../PullRequestWidget';
import { getPullRequestData } from '../../api';

jest.mock('../../api', () => ({
  getPullRequestData: jest.fn(() =>
    Promise.resolve({
      key: '1',
      isPullRequest: true,
      projectKey: 'foo',
      status: { qualityGateStatus: 'ERROR', codeSmells: 12, bugs: 2, vulnerabilities: 1 }
    })
  )
}));
jest.mock('../../utils', () => ({
  getRepoSettingsUrl: jest.fn(() =>
    Promise.resolve(
      'https://bitbucketcloud.org/{}/{repo_uuid}/admin/addon/admin/app-key/repository-config'
    )
  ),
  isRepoAdmin: jest.fn(() => true)
}));

const CONTEXT = { repoUuid: 'repo_uuid', prId: '1' };

beforeEach(() => {
  (getPullRequestData as jest.Mock<any>).mockClear();
});

it('should display correctly', async () => {
  const wrapper = shallow(<PullRequestWidget context={CONTEXT} />);
  expect(wrapper).toMatchSnapshot();
  expect(getPullRequestData).toHaveBeenCalledWith(CONTEXT);

  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('should display a no analysis message', async () => {
  (getPullRequestData as jest.Mock<any>).mockImplementationOnce(() => Promise.reject('Not found'));
  const wrapper = shallow(<PullRequestWidget context={CONTEXT} />);
  expect(getPullRequestData).toHaveBeenCalledWith(CONTEXT);

  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('should display a no issues message', async () => {
  (getPullRequestData as jest.Mock<any>).mockImplementationOnce(() =>
    Promise.resolve({
      key: '1',
      projectKey: 'foo',
      status: { qualityGateStatus: 'OK', codeSmells: 0, bugs: 0, vulnerabilities: 0 }
    })
  );
  const wrapper = shallow(<PullRequestWidget context={CONTEXT} />);
  expect(getPullRequestData).toHaveBeenCalledWith(CONTEXT);

  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});

it('should display projects not bound message', async () => {
  (getPullRequestData as jest.Mock<any>).mockImplementationOnce(
    jest.fn(() => Promise.reject({ errors: [{ msg: 'Repository not bound' }] }))
  );
  const wrapper = shallow(<PullRequestWidget context={CONTEXT} />);
  expect(getPullRequestData).toHaveBeenCalledWith(CONTEXT);

  await new Promise(setImmediate);
  wrapper.update();
  expect(wrapper).toMatchSnapshot();
});
