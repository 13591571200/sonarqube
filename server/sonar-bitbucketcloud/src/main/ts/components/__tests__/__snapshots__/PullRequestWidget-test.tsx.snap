// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a no analysis message 1`] = `
<p
  className="pr-widget empty"
>
  Not analyzed on SonarCloud yet
</p>
`;

exports[`should display a no issues message 1`] = `
<p
  className="pr-widget"
>
  <span
    title="Quality Gate OK"
  >
    <StatusIndicator
      className="little-spacer-left little-spacer-right"
      color="green"
      size="small"
    />
  </span>
  Hooray! SonarCloud analysis didn't find any issue!
</p>
`;

exports[`should display correctly 1`] = `
<div
  className="pr-widget"
>
  <Spinner
    delay={100}
    invertColor={false}
    isCompleting={false}
    onComplete={[Function]}
    size="small"
  />
</div>
`;

exports[`should display correctly 2`] = `
<div
  className="pr-widget"
>
  <div
    className="pr-widget-inner"
  >
    <span
      title="Quality Gate ERROR"
    >
      <StatusIndicator
        className="little-spacer-left little-spacer-right"
        color="red"
        size="small"
      />
    </span>
    <span
      className="little-spacer-right"
    >
      SonarCloud found issues:
    </span>
    <span
      className="little-spacer-right"
    >
      <BugIcon
        className="little-spacer-right"
      />
       
      2
       Bugs
    </span>
    <span
      className="little-spacer-right"
    >
      <VulnerabilityIcon
        className="little-spacer-right"
      />
       
      1
       Vulnerabilities
    </span>
    <span
      className="little-spacer-right"
    >
      <CodeSmellIcon
        className="little-spacer-right"
      />
       
      12
       Code Smells
    </span>
  </div>
  <a
    href="/project/issues?id=foo&pullRequest=1&resolved=false"
    target="_blank"
  >
    See all issues on SonarCloud
  </a>
</div>
`;

exports[`should display projects not bound message 1`] = `
<p
  className="pr-widget empty"
>
  <React.Fragment>
    No link between repository and SonarCloud project. Go to your 
    <a
      href="https://bitbucketcloud.org/{}/{repo_uuid}/admin/addon/admin/app-key/repository-config"
      rel="noopener noreferrer"
      target="_parent"
    >
      repository settings
    </a>
    .
  </React.Fragment>
</p>
`;
